src/turing_machine/machine.py

```diff
@@ -1,41 +1,83 @@
-class TuringMachine:
-    def __init__(self, tape, initial_state, final_states, transition_function):
+from abc import ABC, abstractmethod
+
+class Tape:
+    def __init__(self, tape):
         self.tape = list(tape)
         self.head = 0
-        self.state = initial_state
+
+    def move_left(self):
+        self.head -= 1
+
+    def move_right(self):
+        self.head += 1
+
+    def get_value(self):
+        return self.tape[self.head]
+
+    def set_value(self, val):
+        self.tape[self.head] = val
+
+    def __str__(self):
+        return ''.join(self.tape)
+
+
+class TuringMachine(ABC):
+    def __init__(self, tapes, initial_state, final_states):
+        self.initial_tapes = [Tape(tape) for tape in tapes]
+        self.initial_state = initial_state
         self.final_states = final_states
-        self.transition_function = transition_function
+        self.reset()
+
+    def reset(self):
+        self.tapes = [Tape(tape.tape) for tape in self.initial_tapes]
+        self.state = self.initial_state
         self.match_position = -1
 
+    def get_tapes(self):
+        return [tape.tape for tape in self.tapes]
+
+    @abstractmethod
+    def run_auto(self):
+        pass
+
+    def set_tapes(self, new_values):
+        for tape, value in zip(self.tapes, new_values):
+            tape.set_value(value)
+
+    def move_tapes(self, operations):
+        for tape, operation in zip(self.tapes, operations):
+            if operation == 1:
+                tape.move_right()
+            elif operation == -1:
+                tape.move_left()
+
     def step(self):
-        current_char = self.tape[self.head]
-        if (self.state, current_char) in self.transition_function:
-            next_state, write_char, direction = self.transition_function[(self.state, current_char)]
-            self.tape[self.head] = write_char
-            self.state = next_state
-            if direction == 'R':
-                self.head += 1
-            elif direction == 'L':
-                self.head -= 1
-            if self.state in self.final_states:
-                self.match_position = self.head
-        else:
-            raise Exception("Error")
+        tape_state = self.get_tapes()
+        new_state, new_values, operations = self.run_auto()
+        self.state = new_state
+        self.set_tapes(new_values)
+        self.move_tapes(operations)
+        if self.state in self.final_states:
+            self.match_position = [tape.head for tape in self.tapes]
```

Comment:
`self.match_position` nie powinno być zmienną składową, a metodą klasy `TuringMachine`. Robiąc ją składową duplikujesz przechowywanie tej samej wartości, co może prowadzić do sytuacji kiedy dla tego samego obiektu znajdą się one w innych stanach (czyli de facto problem 'single source of truth').

Tak więc utwórz metodę w klasie `TuringMachine` o nazwie `tape_positions` która będzie zwracała dokładnie wartość przechowywaną w składowej `self.match_position` (a tą składową usuń z kodu).